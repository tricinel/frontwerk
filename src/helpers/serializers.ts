import slash from 'slash';

// Remove the quotes around strings
const unquoteSerializer = {
  print: <T>(val: T): T => val,
  test: <T>(val: T): boolean => typeof val === 'string'
};

// Cconvert windows style file paths to unix
const winPathSerializer = {
  print: (val: string): string => slash(val),
  test: <T>(val: T): boolean => typeof val === 'string' && val.includes('\\')
};

export { unquoteSerializer, winPathSerializer };
