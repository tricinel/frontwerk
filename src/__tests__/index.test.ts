/* eslint-disable global-require */
import jestSerializerPath from 'jest-serializer-path';
import cases from 'jest-in-case';

import { unquoteSerializer, winPathSerializer } from '../helpers/serializers';

expect.addSnapshotSerializer(unquoteSerializer);
expect.addSnapshotSerializer(winPathSerializer);
expect.addSnapshotSerializer(jestSerializerPath);

const testCases = [
  {
    name: 'errors out when using the wrong node version',
    nodeVersion: 'v6.11.0',
    throws: true
  },
  {
    name: 'calls node with the script path and args',
    args: ['test', '--no-watch']
  },
  {
    name: 'throws unknown script',
    args: ['unknown-script'],
    throws: true
  },
  {
    name: 'logs help with no args',
    snapshotLog: true
  },
  {
    name: 'handles SIGKILL',
    args: ['test'],
    signal: 'SIGKILL'
  },
  {
    name: 'handles SIGTERM',
    args: ['test'],
    signal: 'SIGTERM'
  }
];

const testFn = ({
  snapshotLog = false,
  throws = false,
  args = [],
  nodeVersion = 'v10',
  signal = null
}) => {
  const crossSpawn = require('cross-spawn');
  const syncSpy = jest
    .spyOn(crossSpawn, 'sync')
    .mockImplementation(() => ({ signal, status: 0 }));
  const consola = require('consola');
  const exitSpy = jest.spyOn(process, 'exit');
  const logSpy = jest.spyOn(consola, 'info').mockImplementation(() => {});
  const errorSpy = jest.spyOn(consola, 'error').mockImplementation(() => {});
  const { argv: originalArgv } = process;

  delete process.version;
  process.version = nodeVersion;

  try {
    // Tests
    process.argv = ['node', '../', ...args];
    require('../');

    if (typeof signal === 'string') {
      expect(errorSpy).toHaveBeenCalledTimes(1);
      expect(errorSpy.mock.calls).toMatchSnapshot();
    } else if (snapshotLog) {
      expect(logSpy).toHaveBeenCalledTimes(1);
      expect(logSpy.mock.calls).toMatchSnapshot();
    } else {
      expect(syncSpy).toHaveBeenCalledTimes(1);
      const [firstCall] = syncSpy.mock.calls;
      const [script, calledArgs] = firstCall as [string, string[]];
      expect([script, ...calledArgs].join(' ')).toMatchSnapshot();
    }
  } catch (error) {
    if (throws) {
      expect(error.message).toMatchSnapshot();
    } else {
      throw error;
    }
  } finally {
    // We reset everything afterEach
    exitSpy.mockRestore();
    logSpy.mockRestore();
    errorSpy.mockRestore();
    syncSpy.mockRestore();
    process.argv = originalArgv;
    jest.resetModules();
  }
};

cases('index', testFn, testCases);
