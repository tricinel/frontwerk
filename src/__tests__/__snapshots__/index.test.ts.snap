// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index calls node with the script path and args 1`] = `node <PROJECT_ROOT>/src/scripts/test.ts --no-watch`;

exports[`index errors out when using the wrong node version 1`] = `You must use Node version 10 or greater`;

exports[`index handles SIGKILL 1`] = `
Array [
  Array [
    The script "test" failed because the process exited too early.
This probably means the system ran out of memory or someone called \`kill -9\` on the process.,
  ],
]
`;

exports[`index handles SIGTERM 1`] = `
Array [
  Array [
    The script "test" failed because the process exited too early.
Someone might have called \`kill\` or \`killall\`, or the system could be shutting down.,
  ],
]
`;

exports[`index logs help with no args 1`] = `
Array [
  Array [
    
Usage: ../ [script] [--flags]
Available Scripts:
build --bundle: Build your ES6 files into a single file using RollupJS
build: Build your ES6 files using BabelJS
lint: Lint your JS files using ESLint
stylelint: Lint your CSS/SCSS/SASS files using Stylelint
format: Format your JS files using Prettier
test: Test your JS files using Jest
Options:
  All options depend on the script. The args you pass will be forwarded to the respective tool that's being run under the hood.
,
  ],
]
`;

exports[`index throws unknown script 1`] = `Unknown script unknown-script. Please check if you need to update frontwerk.`;
